examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+    top_k : int, required.
+        Integer defining the top number of paths to decode.
+    Returns
+    -------
+    viterbi_path : List[int]
+        The tag indices of the maximum likelihood tag sequence.
+    viterbi_score : float
+        The score of the viterbi path.
+    """
+    sequence_length, num_tags = tag_sequence.size()
+
+    path_scores = []
+    path_indices = []
+    # At the beginning, the maximum number of permutations is 1; therefore, we unsqueeze(0)
+    # to allow for 1 permutation.
+    path_scores.append(tag_sequence[0, :].unsqueeze(0))
+    # assert path_scores[0].size() == (n_permutations, num_tags)
+
+    # Evaluate the scores for all possible paths.
+    for timestep in range(1, sequence_length):
+        # Add pairwise potentials to current scores.
+        # assert path_scores[timestep - 1].size() == (n_permutations, num_tags)
+        summed_potentials = path_scores[timestep -
+                                        1].unsqueeze(2) + transition_matrix
+        summed_potentials = summed_potentials.view(-1, num_tags)
+
+        # Best pairwise potential path score from the previous timestep.
+        max_k = min(summed_potentials.size()[0], top_k)
+        scores, paths = torch.topk(summed_potentials, k=max_k, dim=0)
+        # assert scores.size() == (n_permutations, num_tags)
+        # assert paths.size() == (n_permutations, num_tags)
+
+        scores = tag_sequence[timestep, :] + scores
+        # assert scores.size() == (n_permutations, num_tags)
+        path_scores.append(scores)
+        path_indices.append(paths.squeeze())
+
+    # Construct the most likely sequence backwards.
+    path_scores = path_scores[-1].view(-1)
+    max_k = min(path_scores.size()[0], top_k)
+    viterbi_scores, best_paths = torch.topk(path_scores, k=max_k, dim=0)
+
+    viterbi_paths = []
+    for i in range(max_k):
+
+        viterbi_path = [best_paths[i].item()]
+        for backward_timestep in reversed(path_indices):
+            viterbi_path.append(
+                int(backward_timestep.view(-1)[viterbi_path[-1]]))
+
+        # Reverse the backward path.
+        viterbi_path.reverse()
+
+        # Viterbi paths uses (num_tags * n_permutations) nodes; therefore, we need to modulo.
+        viterbi_path = [j % num_tags for j in viterbi_path]
+        viterbi_paths.append(viterbi_path)
+
+    return viterbi_paths, viterbi_scores
+
+
+def batch_viterbi_decode(tag_sequence: torch.Tensor, transition_matrix: torch.Tensor, top_k: int = 5):
+
+    outputs = []
+    scores = []
+    for i in range(tag_sequence.shape[1]):
+        paths, score = viterbi_decode(tag_sequence[:, i, :], transitions)
+        outputs.append(paths)
+        scores.append(score)
+
+    return torch.tensor(outputs).transpose(0, -1), torch.cat(scores)
+
+
+def top_batch_viterbi_decode(tag_sequence: torch.Tensor):
+    output, _ = batch_viterbi_decode(tag_sequence, transitions, top_k=1)
+    return output[:, 0, :]
+
+
+def levenshtein_distance_list(r, h):
```

Comment:
This seems unused

Reply:
Only one version is needed indeed. The list version is faster than the pytorch version though.